zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterFactory.java

```diff
@@ -917,6 +918,164 @@ private Properties updatePropertiesFromRegisteredInterpreter(Properties properti
     return properties;
   }
 
+  /**
+   * map interpreter ids into noteId
+   * @param noteId note id
+   * @param ids InterpreterSetting id list
+   * @throws IOException
+   */
+  public void setInterpreters(String noteId, List<String> ids) throws IOException {
+    putNoteInterpreterSettingBinding(noteId, ids);
+  }
+
+  public List<String> getInterpreters(String noteId) {
+    return getNoteInterpreterSettingBinding(noteId);
+  }
+
+  public List<InterpreterSetting> getInterpreterSettings(String noteId) {
+    List<String> interpreterSettingIds = getNoteInterpreterSettingBinding(noteId);
+    LinkedList<InterpreterSetting> settings = new LinkedList<InterpreterSetting>();
+    synchronized (interpreterSettingIds) {
+      for (String id : interpreterSettingIds) {
+        InterpreterSetting setting = get(id);
+        if (setting == null) {
+          // interpreter setting is removed from factory. remove id from here, too
+          interpreterSettingIds.remove(id);
+        } else {
+          settings.add(setting);
+        }
+      }
+    }
+    return settings;
+  }
+
+  public void closeNote(String noteId) {
+    // close interpreters in this note session
+    List<InterpreterSetting> settings = getInterpreterSettings(noteId);
+    if (settings == null || settings.size() == 0) {
+      return;
+    }
+
+    System.err.println("close");
+    for (InterpreterSetting setting : settings) {
+      removeInterpretersForNote(setting, noteId);
+    }
+  }
+
+  private String getInterpreterInstanceKey(String noteId, InterpreterSetting setting) {
+    if (setting.getOption().isExistingProcess()) {
+      return Constants.EXISTING_PROCESS;
+    } else if (setting.getOption().isPerNoteSession() || setting.getOption().isPerNoteProcess()) {
+      return noteId;
+    } else {
+      return SHARED_SESSION;
+    }
+  }
+
+  private List<Interpreter> createOrGetInterpreterList(String noteId, InterpreterSetting setting) {
+    InterpreterGroup interpreterGroup =
+        setting.getInterpreterGroup(noteId);
+    synchronized (interpreterGroup) {
+      String key = getInterpreterInstanceKey(noteId, setting);
+      if (!interpreterGroup.containsKey(key)) {
+        createInterpretersForNote(setting, noteId, key);
+      }
+      return interpreterGroup.get(getInterpreterInstanceKey(noteId, setting));
+    }
+  }
+
+  private com.google.common.base.Optional<InterpreterSetting>
+  getDefaultInterpreterSetting(List<InterpreterSetting> settings) {
+    if (settings == null || settings.isEmpty()) {
+      return com.google.common.base.Optional.absent();
+    }
+    return com.google.common.base.Optional.of(settings.get(0));
+  }
+
+  public com.google.common.base.Optional<InterpreterSetting> getDefaultInterpreterSetting(String noteId) {
```

Comment:
This is because the version of JDK installed in my mac is 1.8.x and I just add `import java.util.*`, then I'll fix it by each packages instead of `*`after passing CI.